File: mekong-tunnel/src/mekong_tunnel/find_mekong.py
"""
Locate the mekong binary on the current system.
No external dependencies — stdlib only.
"""

import os
import shutil
import sys


def find_mekong() -> 'str | None':
    """
    Search for the mekong binary in common locations.

    Order:
      1. shutil.which('mekong')  — skipped if the path contains 'mekong_tunnel'
         (that would resolve to this package's own console-script wrapper)
      2. ~/.local/bin/mekong
      3. /usr/local/bin/mekong
      4. /usr/bin/mekong
      5. Windows extras: same paths with .exe + %USERPROFILE%\\AppData\\Local\\mekong.exe

    Returns the absolute path string, or None if not found.
    """
    is_windows = sys.platform.startswith('win')
    suffixes = ['.exe', ''] if is_windows else ['']

    # --- 1. PATH lookup via shutil.which ---
    for suffix in suffixes:
        candidate = shutil.which('mekong' + suffix)
        if candidate and 'mekong_tunnel' not in candidate:
            return candidate

    # --- 2-4. Well-known fixed paths ---
    home = os.path.expanduser('~')

    # /usr/local/bin first — macOS default install location (always in VS Code PATH)
    fixed_paths = [
        '/usr/local/bin/mekong',
        os.path.join(home, '.local', 'bin', 'mekong'),
        os.path.join(home, 'bin', 'mekong'),
        '/usr/bin/mekong',
        '/opt/homebrew/bin/mekong',
    ]

    if is_windows:
        local = os.environ.get('LOCALAPPDATA', os.path.join(home, 'AppData', 'Local'))
        userprofile = os.environ.get('USERPROFILE', home)
        fixed_paths += [
            os.path.join(local, 'Programs', 'mekong', 'mekong.exe'),
            os.path.join(local, 'mekong.exe'),
            os.path.join(home, '.local', 'bin', 'mekong.exe'),
            os.path.join(userprofile, 'AppData', 'Local', 'Programs', 'mekong', 'mekong.exe'),
        ]

    for path in fixed_paths:
        if os.path.isfile(path) and os.access(path, os.X_OK):
            return path

    return None

File: mekong-node-sdk/lib/find-mekong.js
'use strict';

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');

/**
 * Locate the mekong binary.
 * Returns an absolute path string or null if not found.
 */
function findMekong() {
  const isWindows = process.platform === 'win32';

  // 1. Try `which mekong` (or `where mekong` on Windows)
  try {
    const cmd = isWindows ? 'where mekong' : 'which mekong';
    const result = execSync(cmd, { stdio: ['ignore', 'pipe', 'ignore'] })
      .toString()
      .trim()
      .split('\n')[0]
      .trim();
    if (result && fs.existsSync(result)) return result;
  } catch (_) {
    // not on PATH
  }

  // 2. Common Unix paths — /usr/local/bin first (macOS default install location)
  if (!isWindows) {
    const unixPaths = [
      '/usr/local/bin/mekong',
      path.join(os.homedir(), '.local', 'bin', 'mekong'),
      path.join(os.homedir(), 'bin', 'mekong'),
      '/usr/bin/mekong',
      '/opt/homebrew/bin/mekong',
    ];
    for (const p of unixPaths) {
      if (fs.existsSync(p)) return p;
    }
  }

  // 3. Windows paths — %LOCALAPPDATA%\Programs\mekong first (default install dir)
  if (isWindows) {
    const local = process.env.LOCALAPPDATA || path.join(os.homedir(), 'AppData', 'Local');
    const winPaths = [
      path.join(local, 'Programs', 'mekong', 'mekong.exe'),
      path.join(local, 'mekong.exe'),
      path.join(os.homedir(), '.local', 'bin', 'mekong.exe'),
      'C:\\Program Files\\mekong\\mekong.exe',
    ];
    for (const p of winPaths) {
      if (fs.existsSync(p)) return p;
    }

    // Also try without .exe
    const winPathsNoExt = [
      path.join(local, 'Programs', 'mekong', 'mekong'),
      path.join(os.homedir(), '.local', 'bin', 'mekong'),
    ];
    for (const p of winPathsNoExt) {
      if (fs.existsSync(p)) return p;
    }
  }

  return null;
}

module.exports = { findMekong };
